sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala

```diff
@@ -1556,6 +1556,20 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
       df.where($"city".contains(new java.lang.Character('A'))),
       Seq(Row("Amsterdam")))
   }
+
+  test("SPARK-25942: typed aggregation on primitive type") {
+    val ds = Seq(1, 2, 3).toDS()
+
+    val agg = ds.groupByKey(_ >= 2)
+      .agg(sum("value").as[Long], sum($"value" + 1).as[Long])
```

Comment:
I think we should not make decisions for users. For untyped APIs, users can refer the grouping columns in the aggregate expressions, I think the typed APIs should be same.

For this particular case, currrently spark allows grouping columns inside aggregate functions, so the `value` here is indeed ambiguous. There is nothing we can do, but fail and ask users to add alias.

BTW, we should check other databases and see if "grouping columns inside aggregate functions" should be allowed,

Reply:
I think typed groupByKey is somehow different to untyped groupBy API.

For untyped groupBy API, the grouping attributes are from original relation and visible to users before grouping. It is reasonable to access grouping attributes in aggregate expressions for untype groupBy.

The key attributes of typed groupByKey API are appended by users in addition to original data attributes. The key attributes are from typed user-provided function. It can be primitive type or complex type like case class.

It makes difficulties to have general and reasonable access to key attributes in aggregate expressions for typed groupByKey API. For example, for case class type, key attributes are flatten alongside original data attributes. So aggregate expressions should access single typed key or flatten attributes? In other words, if key object is `(Int, String)`, how aggregate expressions use it?

From the detail of `aggUntyped`, it also shows that key attributes should not be accessed in aggregate expressions in typed groupByKey API. In `aggUntyped`, we call `withInputType` to bind value encoders on data attributes. So typed aggregate expressions can't access key attributes too.

Reply:
Btw, the definition of `agg` functions in `KeyValueGroupedDataset` look like:

```scala
def agg[U1, U2](col1: TypedColumn[V, U1], col2: TypedColumn[V, U2]): Dataset[(K, U1, U2)]
```

So the inputs to `agg` are `TypedColumn` with `V` as input type.

Reply:
ah i see your point.

For untyped API, `df.groupBy...agg...` produces one plan node, but the typed API `df.groupByKey...agg...` produces 2 plan nodes.

makes sense to me.

Reply:
looking at `KeyValueGroupedDataset.dataAttributes`, seems we do want to resolve things based on the plan below `AppendColumns`. And in `KeyValueGroupedDataset#aggUntyped`, we do resolve expressions based on `dataAttributes`. Do you why it doesn't work?

Reply:
For `TypedAggregateExpression` input, `TypedColumn.withInputType` will create `UnresolvedDeserializer` to resolve on `dataAttributes`.

But for non `TypedAggregateExpression`, `withInputType` actually doesn't do anything. So such non typed aggregate expressions, we still resolve expressions on `dataAttributes` + `groupingAttributes`.

Reply:
ah i see. Can you add a comment in `TypedColumn.withInputType` and say that untyped normal aggregate expressions are handled in the analyzer directly?

Reply:
`TypedAggregateExpression.withInputInfo` needs the `UnresolvedDeserializer` which depends on input encoder and input attributes. In analyzer, we can't have such inputs.